sources/bootstrap-commands/src/main.rs

```diff
@@ -0,0 +1,370 @@
+/*!
+# Bootstrap commands
+
+`bootstrap-commands` ensures that bootstrap commands are executed as defined in the system
+settings. It is called by `bootstrap-commands.service` which runs prior to the execution of
+`bootstrap-containers`.
+
+Each bootstrap command is a set of Bottlerocket API commands. The settings are first rendered
+into a config file. Then, the system is configured by going through all the bootstrap commands
+in lexicographical order and running all the commands inside it.
+
+## Example:
+Given a bootstrap command called `001-test-bootstrap-commands` with the following configuration:
+
+```toml
+[settings.bootstrap-commands.001-test-bootstrap-commands]
+commands = [[ "apiclient", "set", "motd=helloworld"]]
+essential = true
+mode = "always"
+```
+This would set `/etc/motd` to "helloworld".
+
+# Additional Information:
+Certain valid `apiclient` commands that work in a session may fail in `bootstrap-commands`
+due to relevant services not running at the time of the launch of the systemd service.
+
+## Example:
+```toml
+[settings.bootstrap-commands.001-test-bootstrap-commands]
+commands = [[ "apiclient", "exec", "admin", "ls"]]
+essential = true
+mode = "always"
+```
+This command fails because `bootstrap-commands.service` which calls this binary is launched
+prior to `preconfigured.target` while `host-containers@.service` which is a requirement for
+running "exec" commands are launched after preconfigured.target.
+*/
+
+use log::info;
+use serde::Deserialize;
+use simplelog::{Config as LogConfig, LevelFilter, SimpleLogger};
+use snafu::{ensure, OptionExt, ResultExt};
+use std::collections::BTreeMap;
+use std::env;
+use std::ffi::OsStr;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::process::{self, Command};
+use std::str::FromStr;
+
+use bottlerocket_modeled_types::{ApiclientCommand, BootstrapMode, Identifier};
+
+const DEFAULT_CONFIG_PATH: &str = "/etc/bootstrap-commands/bootstrap-commands.toml";
+
+#[derive(Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct BootstrapCommandConfig {
+    #[serde(default)]
+    bootstrap_commands: BTreeMap<Identifier, BootstrapCommand>,
+}
+
+impl BootstrapCommandConfig {
+    // Gets an iterator for bootstrap_commands, sorted in lexicographical order of their names.
+    fn iter(self) -> impl Iterator<Item = (Identifier, BootstrapCommand)> {
+        self.bootstrap_commands.into_iter()
+    }
+}
```

Comment:
I think this is fine, but as a Rust tip, there's the [`IntoIter`](https://doc.rust-lang.org/stable/std/iter/trait.IntoIterator.html) in `std` that you can implement for this behavior, which can make calling other APIs that expect `IntoIter` ergonomic.